gh-135751: traceback: add recent_first and show_source_lines (#135752)
python/cpython: Doc/library/traceback.rst

   :term:`file <file object>` or :term:`file-like object` to
   receive the output.

>    If *show_lines* is false, source code lines are not included in the output.

Reviewer: Flip the double negative? 
```suggestion
   If *show_lines* is true, source code lines are included in the output.
```

Reviewer 2: If flipped, it might be helpful to mention that true is the default
```suggestion
   If *show_lines* is true (the default), source code lines will be included in the output.
```

   If *recent_first* is true, the most recent stack trace entries are printed
   first, otherwise the oldest entries are printed first. The default is false.

   .. note::
      ``recent_first=True`` is useful for showing stack traces in places where

>       people see the top of the stack trace first, such as in a web browser.

Reviewer 2: I'm not sure this needs the loudness of a `.. note` block. Merging this into the previous paragraph would be sufficient IMO

   :attr:`~BaseException.__cause__` or :attr:`~BaseException.__context__`
   attributes of the exception) will be
   printed as well, like the interpreter itself does when printing an unhandled
   exception.  If *show_lines* is ``False``, source code lines are not included

>    in the output.

Author: ```suggestion
   exception.
```

   printed as well, like the interpreter itself does when printing an unhandled
   exception.

>    If *show_lines* is false, source code lines are not included in the output.

Reviewer: ```suggestion
   If *show_lines* is true, source code lines are included in the output.
```

   This is a shorthand for ``print_exception(sys.exception(), limit=limit, file=file,
   chain=chain, show_lines=show_lines, recent_first=recent_first)``.

   .. versionchanged:: next
      Added *show_lines* and *recent_first* parameters.

.. function:: print_last(limit=None, file=None, chain=True, *, show_lines=True, recent_first=False)

   This is a shorthand for ``print_exception(sys.last_exc, limit=limit, file=file,
   chain=chain, show_lines=show_lines, recent_first=recent_first)``.
   In general it will work only after an exception has reached an interactive
   prompt (see :data:`sys.last_exc`).

   .. versionchanged:: next
      Added *show_lines* and *recent_first* parameters.

.. function:: print_stack(f=None, limit=None, file=None, *, show_lines=True, recent_first=False)

   Print up to *limit* stack trace entries (starting from the invocation
   point) if *limit* is positive.  Otherwise, print the last ``abs(limit)``
   entries.  If *limit* is omitted or ``None``, all entries are printed.
   The optional *f* argument can be used to specify an alternate
   :ref:`stack frame <frame-objects>`
   to start.  The optional *file* argument has the same meaning as for

>    :func:`print_tb`.  If *show_lines* is ``False``, source code lines are

Reviewer: ```suggestion
   :func:`print_tb`.  If *show_lines* is true, source code lines are
```

>    not included in the output.

Reviewer: ```suggestion
   included in the output.
```

   arguments have the same meaning as for :func:`print_stack`.

.. function:: print_list(extracted_list, file=None, *, show_lines=True)

   Print the list of tuples as returned by :func:`extract_tb` or
   :func:`extract_stack` as a formatted stack trace to the given file.
   If *file* is ``None``, the output is written to :data:`sys.stderr`.

>    If *show_lines* is ``False``, source code lines are not included in the output.

Reviewer: ```suggestion
   If *show_lines* is true, source code lines are included in the output.
```

   .. versionchanged:: next

>       Added *show_lines* parameter.

Reviewer 2: nit: some of new version changed directives use the style "Added X parameter" while others use "Added *the* X parameter". Both styles are used elsewhere in the docs, but it would probably be better to use one consistently here

.. function:: format_list(extracted_list, *, show_lines=True)

   Given a list of tuples or :class:`FrameSummary` objects as returned by
   :func:`extract_tb` or :func:`extract_stack`, return a list of strings ready
   for printing.  Each string in the resulting list corresponds to the item with
   the same index in the argument list.  Each string ends in a newline; the
   strings may contain internal newlines as well, for those items whose source

>    text line is not ``None``.  If *show_lines* is ``False``, source code lines

Reviewer: ```suggestion
   text line is not ``None``.  If *show_lines* is ``True``, source code lines
```

>    are not included in the output.

Reviewer: ```suggestion
   are included in the output.
```

      *show_group* parameter was added.

.. function:: format_exception(exc, /[, value, tb], limit=None, chain=True, *, show_lines=True, recent_first=False, show_group=False)

   Format a stack trace and the exception information.  The arguments  have the
   same meaning as the corresponding arguments to :func:`print_exception`.  The
   return value is a list of strings, each ending in a newline and some
   containing internal newlines.  When these lines are concatenated and printed,
   exactly the same text is printed as does :func:`print_exception`.

>    If *show_lines* is false, source code lines are not included in the output.

Reviewer: ```suggestion
   If *show_lines* is true, source code lines are included in the output.
```

   If *recent_first* is true, the most recent stack trace entries are printed
   first, otherwise the oldest entries are printed first. The default is false.

   .. versionchanged:: 3.5
      The *etype* argument is ignored and inferred from the type of *value*.

   .. versionchanged:: 3.10
      This function's behavior and signature were modified to match
      :func:`print_exception`.

   .. versionchanged:: next
      Added *show_lines* and *recent_first* parameters.

.. function:: format_exc(limit=None, chain=True, *, show_lines=True, recent_first=False)

   This is like ``print_exc(limit)`` but returns a string instead of printing to
   a file.

>    If *show_lines* is false, source code lines are not included in the output.

Reviewer: ```suggestion
   If *show_lines* is true, source code lines are included in the output.
```

      For syntax errors - the compiler error message.

   .. classmethod:: from_exception(exc, *, limit=None, lookup_lines=False, capture_locals=False)

      Capture an exception for later rendering. *limit*, *lookup_lines* and
      *capture_locals* are as for the :class:`StackSummary` class.

      Note that when locals are captured, they are also shown in the traceback.

      .. versionchanged:: next
         Changed *lookup_lines* default to ``False`` to avoid overhead when
         formatting exceptions with ``show_lines=False``.

   .. method::  print(*, file=None, chain=True, show_lines=True, recent_first=False)

      Print to *file* (default ``sys.stderr``) the exception information returned by
      :meth:`format`.

>       If *show_lines* is false, source code lines from being included in the output.

Reviewer: ```suggestion
      If *show_lines* is true, source code lines are included in the output.
```

      some containing internal newlines. :func:`~traceback.print_exception`
      is a wrapper around this method which just prints the lines to a file.

>       If *show_lines* is false, source code lines from being included in the output.

Reviewer: ```suggestion
      If *show_lines* is true, source code lines are included in the output.
```

      repetitions are shown, followed by a summary line stating the exact
      number of further repetitions.

>       The keyword argument *show_lines*, if ``False``, prevents source code
>       lines from being included in the output.

Reviewer 2: ```suggestion
      If *show_lines* is set to ``False``, source code lines will not be
      included in the output.
```

Reviewer: Or flip to remove double negative.

>       The keyword argument *show_lines*, if ``False``, prevents source code
>       lines from being included in the output.

Author: ```suggestion
      If *show_lines* is true, includes source code lines in the output.
```

      .. versionchanged:: 3.6
         Long sequences of repeated frames are now abbreviated.

      .. versionchanged:: next
         Added the *show_lines* parameter.

   .. method:: format_frame_summary(frame_summary, *, show_lines=True, **kwargs)

      Returns a string for printing one of the :ref:`frames <frame-objects>`
      involved in the stack.
      This method is called for each :class:`FrameSummary` object to be
      printed by :meth:`StackSummary.format`. If it returns ``None``, the
      frame is omitted from the output.

>       The keyword argument *show_lines*, if ``False``, prevents source code

Reviewer: ```suggestion
      The keyword argument *show_lines*, if ``True``, includes source code
```

>       lines from being included in the output.

Reviewer: ```suggestion
      lines in the output.
```
